File: backend/app/schemas/availability.py
from datetime import date, time

from pydantic import BaseModel


class RuleCreate(BaseModel):
    weekday: int
    start_time: time
    end_time: time
    effective_from: date | None = None
    effective_until: date | None = None


class RuleOut(RuleCreate):
    id: int
    effective_from: date
    model_config = {"from_attributes": True}


class DayWindow(BaseModel):
    start_time: time
    end_time: time


class DayOverrideSave(BaseModel):
    """Полное описание дня: closed=true ИЛИ список окон. Пустой список окон при closed=false = вернуть день к правилам (удалить overrides)."""
    closed: bool = False
    windows: list[DayWindow] = []


class DayOverrideOut(BaseModel):
    date: date
    overridden: bool
    closed: bool
    windows: list[DayWindow]


class WindowOut(BaseModel):
    date: date
    start_time: time
    end_time: time


class AvailabilitySaveResult(BaseModel):
    conflicts: list[dict]  # [{lesson_id, starts_at, student_name}] — занятия вне новой доступности
    rejected_bookings: int  # сколько заявок авто-отклонено

File: web/src/api/availability.ts
import { apiClient } from './client';

export interface AvailabilityRule {
  id: number;
  weekday: number; // 0=Пн
  start_time: string; // "14:00:00"
  end_time: string;
  effective_from: string;
  effective_until: string | null;
}

export interface DayWindow {
  start_time: string;
  end_time: string;
}

export interface DayOverride {
  date: string;
  overridden: boolean;
  closed: boolean;
  windows: DayWindow[];
}

export interface ComputedWindow {
  date: string;
  start_time: string;
  end_time: string;
}

export interface SaveResult {
  conflicts: { lesson_id: number; starts_at: string; student_name: string | null }[];
  rejected_bookings: number;
}

export const listRules = async (): Promise<AvailabilityRule[]> => {
  const response = await apiClient.get('/availability/rules');
  return response.data;
};

export const replaceRules = async (
  rules: { weekday: number; start_time: string; end_time: string }[]
): Promise<SaveResult> => {
  const response = await apiClient.put('/availability/rules', rules);
  return response.data;
};

export const getDay = async (day: string): Promise<DayOverride> => {
  const response = await apiClient.get(`/availability/day/${day}`);
  return response.data;
};

export const saveDay = async (
  day: string,
  payload: { closed: boolean; windows: DayWindow[] }
): Promise<SaveResult> => {
  const response = await apiClient.put(`/availability/day/${day}`, payload);
  return response.data;
};

export const getWindows = async (from: string, to: string): Promise<ComputedWindow[]> => {
  const response = await apiClient.get('/availability/windows', {
    params: { date_from: from, date_to: to },
  });
  return response.data;
};
